prefq_documentation.md

```diff
@@ -0,0 +1,96 @@
+3. **Server**: Receives queries, stores them locally & serves them, whenever requested. Only once all Feedback has been evaluated, returns the Feedback to the Query Client. 
+
+Seperating the Query Client from the Server allows for more flexibility. For instance, you can send Queries to the server from your laptop, then disconnect from the web, and still serve videos to the Feedback Client. Additionally, the Query Client can generate videos and train the feedback model using any desired programming language. Communication happens using an html-request interface. 
+
+## Development Stage Server (local)
+
+For a quick first impression - or if you'd like to adjust the server script to your individual needs - it is not necessary to host a remote server. Instead you can simply run the provided server script on localhost. This behavior is specified within the last lines of `server.py`:
+
+```python
+if __name__ == "__main__":
+    before_first_request()
+    app.run(host="localhost", port=5000, debug=True)
+```
+
+### Server
+Start your own local server:
+1. Clone the repository: `git clone https://github.com/timokau/prefq.git`
+2. Install poetry: `curl -sSL https://install.python-poetry.org | python3 -`
+3. Add entry to .bashrc:<br>   `export PATH="/root/.local/bin:$PATH"`<br>   `echo 'export PATH="/root/.local/bin:$PATH"'`<br>   `~/.bashrc`
+4. Start the server: `scripts/run_server.sh`
+
+### Query Client
+Send queries to your local server:
+1. Open a new terminal
+2. *(a)* Send static queries: `scripts/run_static_queries.sh`<br>*(b)* Send dynamic queries: `scripts/run_imitation_example.sh`
+
+Option (a):
+- shows the required http-request structure for sending queries & receiving feedback
+- does not learn new behavior
+
+Option (b):
+- integrates our PrefQ-Server into [imitation](https://pypi.org/project/imitation/), a library based on [gymnasium](https://gymnasium.farama.org/index.html)
+- capable of learning new behavior
+
+### Feedback Client
+Evaluate queries:
+1. Visit the web interface: `http://localhost:5000` (*using your favorite web browser*)
+2. Select your preferred video: use left/right arrow keys or click the corresponding button
+
+Using such a development server holds several eminent advantages
+1. Changes within your project folder will be recognized. As soon as you save changes, the server will reboot automatically and you can instantly test and verify new behavior.
+2. Valuable debug information about incoming http-requests will be provided.
+3. You can develop your own server without being connected to the web.
+
+
+## Production Stage Server (remote)
+
+For remote feedback collection it is also possible to run a development stage server remotely. To do so you can replace `"localhost"` with the default route `"0.0.0.0"`. This will allow incoming http requests from any IP address.
+
+However, for security reasons it is recommended to use a WSGI server instead. This can be achieved by:
```

Comment:
In what way does this aid security?

Reply:
I did not really dig deeper into this, however, in the flask documentation, it is mentioned, that the development server provided by Werkzeug is convienient yet "not designed to be particularly efficient, stable, or secure" [[1]](https://flask.palletsprojects.com/en/2.3.x/tutorial/deploy/)